c/tskit/tables.c

```diff
@@ -4518,31 +4481,86 @@ table_sorter_sort_mutations(table_sorter_t *self)
     return ret;
 }
 
-static int
-table_sorter_run(table_sorter_t *self, tsk_size_t edge_start)
+int
+tsk_table_sorter_run(tsk_table_sorter_t *self, tsk_bookmark_t *start)
 {
     int ret = 0;
+    tsk_size_t edge_start = 0;
 
-    ret = table_sorter_sort_edges(self, edge_start);
+    if (start != NULL) {
+        if (start->edges > self->tables->edges.num_rows) {
+            ret = TSK_ERR_EDGE_OUT_OF_BOUNDS;
+            goto out;
+        }
+        edge_start = start->edges;
+
+        /* For now, if migrations, sites or mutations are non-zero we get an error.
+         * This should be fixed: https://github.com/tskit-dev/tskit/issues/101
+         */
+        if (start->migrations != 0 || start->sites != 0 || start->mutations != 0) {
+            ret = TSK_ERR_SORT_OFFSET_NOT_SUPPORTED;
+            goto out;
+        }
+    }
+    /* The indexes will be invalidated, so drop them */
+    ret = tsk_table_collection_drop_index(self->tables, 0);
     if (ret != 0) {
         goto out;
     }
-    ret = table_sorter_sort_sites(self);
+    if (self->sort_edges != NULL) {
+        ret = self->sort_edges(self, edge_start);
+        if (ret != 0) {
+            goto out;
+        }
+    }
+    ret = tsk_table_sorter_sort_sites(self);
     if (ret != 0) {
         goto out;
     }
-    ret = table_sorter_sort_mutations(self);
+    ret = tsk_table_sorter_sort_mutations(self);
```

Comment:
IIRC, it was decided that we would allow replacement of the sorting function only for edge sorting, because that was seen as taking the bulk of the total time, right?  At this stage, does that still feel like the right decision, or should the same kind of changes be made for sorting the other tables too, whether for maximal performance or just for parallelism in the APIs?  I have no opinion on this, just asking.  :->

Reply:
Basically the door is open to adding replacement functions for sites and mutations, if we ever want to do it. We can imagine adding function pointers like ``sort_edges`` to the struct, which will for API forward compatible. I didn't do it here because documenting this stuff precisely is quite time consuming, and I doubt there's much actual point for the reasons you point out.